File: crates/jvmjs-compiler/src/stdlib/theater.java
// Bundled clean-room org.code.theater + org.code.media (Code.org CSA).
// Auto-injected when a source imports org.code.theater or org.code.media.
// Phase 1 (headless): Scene records draw commands; Theater.playScenes writes
// them to the VFS file "theater.log" for inspection. Canvas is 400x400 and
// Color values match the real library; the pixel/GIF renderer is Phase 2.
// Only Scene/Theater/Color/Image/Pixel/Font/FontStyle/Instrument are public.

enum Font { MONO, SANS }
enum FontStyle { NORMAL, BOLD, ITALIC }
enum Instrument { PIANO, BASS }

class Color {
  int red, green, blue;
  public Color(int red, int green, int blue) { this.red = red; this.green = green; this.blue = blue; }
  public Color(Color c) { this.red = c.red; this.green = c.green; this.blue = c.blue; }
  public Color(String name) {
    int[] rgb = __resolve(name);
    if (rgb == null) throw new IllegalArgumentException("Invalid color " + name);
    this.red = rgb[0]; this.green = rgb[1]; this.blue = rgb[2];
  }
  public int getRed() { return red; }
  public int getGreen() { return green; }
  public int getBlue() { return blue; }
  public static Color copyWithRed(Color c, int v) { return new Color(v, c.green, c.blue); }
  public static Color copyWithGreen(Color c, int v) { return new Color(c.red, v, c.blue); }
  public static Color copyWithBlue(Color c, int v) { return new Color(c.red, c.green, v); }
  public String toString() { return red + " " + green + " " + blue; }
  static int[] __resolve(String s) {
    String n = s.toUpperCase();
    if (n.length() == 7 && n.charAt(0) == '#') {
      return new int[]{ __hex(n.substring(1, 3)), __hex(n.substring(3, 5)), __hex(n.substring(5, 7)) };
    }
    if (n.equals("WHITE")) return new int[]{255,255,255};
    if (n.equals("SILVER")) return new int[]{192,192,192};
    if (n.equals("GRAY")) return new int[]{128,128,128};
    if (n.equals("BLACK")) return new int[]{0,0,0};
    if (n.equals("RED")) return new int[]{255,0,0};
    if (n.equals("MAROON")) return new int[]{128,0,0};
    if (n.equals("YELLOW")) return new int[]{255,255,0};
    if (n.equals("OLIVE")) return new int[]{128,128,0};
    if (n.equals("LIME")) return new int[]{0,256,0};
    if (n.equals("GREEN")) return new int[]{0,128,0};
    if (n.equals("AQUA")) return new int[]{0,255,255};
    if (n.equals("TEAL")) return new int[]{0,128,128};
    if (n.equals("BLUE")) return new int[]{0,0,255};
    if (n.equals("NAVY")) return new int[]{0,0,128};
    if (n.equals("FUCHSIA")) return new int[]{255,0,255};
    if (n.equals("PURPLE")) return new int[]{128,0,128};
    if (n.equals("PINK")) return new int[]{255,192,203};
    if (n.equals("ORANGE")) return new int[]{255,165,0};
    if (n.equals("GOLD")) return new int[]{255,215,0};
    if (n.equals("BROWN")) return new int[]{165,42,42};
    if (n.equals("CHOCOLATE")) return new int[]{210,105,30};
    if (n.equals("TAN")) return new int[]{210,180,140};
    if (n.equals("TURQUOISE")) return new int[]{64,224,208};
    if (n.equals("INDIGO")) return new int[]{75,0,130};
    if (n.equals("VIOLET")) return new int[]{238,130,238};
    if (n.equals("BEIGE")) return new int[]{245,245,220};
    if (n.equals("IVORY")) return new int[]{255,255,240};
    return null;
  }
  static int __hex(String h) {
    return Integer.parseInt(h, 16);
  }
  public static final Color WHITE = new Color(255,255,255);
  public static final Color SILVER = new Color(192,192,192);
  public static final Color GRAY = new Color(128,128,128);
  public static final Color BLACK = new Color(0,0,0);
  public static final Color RED = new Color(255,0,0);
  public static final Color MAROON = new Color(128,0,0);
  public static final Color YELLOW = new Color(255,255,0);
  public static final Color OLIVE = new Color(128,128,0);
  public static final Color LIME = new Color(0,256,0);
  public static final Color GREEN = new Color(0,128,0);
  public static final Color AQUA = new Color(0,255,255);
  public static final Color TEAL = new Color(0,128,128);
  public static final Color BLUE = new Color(0,0,255);
  public static final Color NAVY = new Color(0,0,128);
  public static final Color FUCHSIA = new Color(255,0,255);
  public static final Color PURPLE = new Color(128,0,128);
  public static final Color PINK = new Color(255,192,203);
  public static final Color ORANGE = new Color(255,165,0);
  public static final Color GOLD = new Color(255,215,0);
  public static final Color BROWN = new Color(165,42,42);
  public static final Color CHOCOLATE = new Color(210,105,30);
  public static final Color TAN = new Color(210,180,140);
  public static final Color TURQUOISE = new Color(64,224,208);
  public static final Color INDIGO = new Color(75,0,130);
  public static final Color VIOLET = new Color(238,130,238);
  public static final Color BEIGE = new Color(245,245,220);
  public static final Color IVORY = new Color(255,255,240);
}

class Pixel {
  Image image; int x, y;
  Pixel(Image image, int x, int y) { this.image = image; this.x = x; this.y = y; }
  public int getX() { return x; }
  public int getY() { return y; }
  public Color getColor() { return image.__get(x, y); }
  public void setColor(Color c) { image.__set(x, y, c); }
  public int getRed() { return getColor().getRed(); }
  public int getGreen() { return getColor().getGreen(); }
  public int getBlue() { return getColor().getBlue(); }
  public void setRed(int v) { Color c = getColor(); image.__set(x, y, new Color(v, c.green, c.blue)); }
  public void setGreen(int v) { Color c = getColor(); image.__set(x, y, new Color(c.red, v, c.blue)); }
  public void setBlue(int v) { Color c = getColor(); image.__set(x, y, new Color(c.red, c.green, v)); }
}

class Image {
  int width, height;
  int[][] r; int[][] g; int[][] b;
  public Image(int width, int height) {
    this.width = width; this.height = height;
    r = new int[width][height]; g = new int[width][height]; b = new int[width][height];
  }
  public Image(String filename) { this(100, 100); }
  public Image(Image src) {
    this.width = src.width; this.height = src.height;
    r = new int[width][height]; g = new int[width][height]; b = new int[width][height];
    for (int i = 0; i < width; i++) for (int j = 0; j < height; j++) { r[i][j]=src.r[i][j]; g[i][j]=src.g[i][j]; b[i][j]=src.b[i][j]; }
  }
  public int getWidth() { return width; }
  public int getHeight() { return height; }
  public Pixel getPixel(int x, int y) { return new Pixel(this, x, y); }
  public void setPixel(int x, int y, Color c) { __set(x, y, c); }
  public void clear(Color c) { for (int i=0;i<width;i++) for(int j=0;j<height;j++) __set(i,j,c); }
  Color __get(int x, int y) { return new Color(r[x][y], g[x][y], b[x][y]); }
  void __set(int x, int y, Color c) { r[x][y]=c.red; g[x][y]=c.green; b[x][y]=c.blue; }
}

class Scene {
  java.util.ArrayList<String> __cmds = new java.util.ArrayList<String>();
  public Scene() {}
  public final int getWidth() { return 400; }
  public final int getHeight() { return 400; }
  java.util.ArrayList<String> __commands() { return __cmds; }
  private String __c(Color c) { return c.getRed() + " " + c.getGreen() + " " + c.getBlue(); }
  public final void clear(String color) { __cmds.add("clear " + __c(new Color(color))); }
  public final void clear(Color color) { __cmds.add("clear " + __c(color)); }
  public final void pause(double seconds) { __cmds.add("pause " + Math.max(seconds, 0.1)); }
  public final void playNote(int note, double seconds) { __cmds.add("note PIANO " + note + " " + seconds); }
  public final void playNote(Instrument inst, int note, double seconds) { __cmds.add("note " + inst + " " + note + " " + seconds); }
  public final void playNoteAndPause(int note, double seconds) { playNote(note, seconds); pause(seconds); }
  public final void playNoteAndPause(Instrument inst, int note, double seconds) { playNote(inst, note, seconds); pause(seconds); }
  public final void playSound(String filename) { __cmds.add("sound " + filename); }
  public final void playSound(double[] sound) { __cmds.add("sound " + sound.length + " samples"); }
  public final void setStrokeWidth(double width) { __cmds.add("strokeWidth " + width); }
  public final void setStrokeColor(String color) { __cmds.add("strokeColor " + __c(new Color(color))); }
  public final void setStrokeColor(Color color) { __cmds.add("strokeColor " + __c(color)); }
  public final void setFillColor(String color) { __cmds.add("fillColor " + __c(new Color(color))); }
  public final void setFillColor(Color color) { __cmds.add("fillColor " + __c(color)); }
  public final void removeStrokeColor() { __cmds.add("strokeColor none"); }
  public final void removeFillColor() { __cmds.add("fillColor none"); }
  public final void setTextColor(String color) { __cmds.add("textColor " + __c(new Color(color))); }
  public final void setTextColor(Color color) { __cmds.add("textColor " + __c(color)); }
  public final void setTextHeight(int height) { __cmds.add("textHeight " + height); }
  public final void setTextStyle(Font font, FontStyle style) { __cmds.add("textStyle " + font + " " + style); }
  public final void drawText(String text, int x, int y) { __cmds.add("text \"" + text + "\" " + x + " " + y + " 0.0"); }
  public final void drawText(String text, int x, int y, double rotation) { __cmds.add("text \"" + text + "\" " + x + " " + y + " " + rotation); }
  public final void drawLine(int sx, int sy, int ex, int ey) { __cmds.add("line " + sx + " " + sy + " " + ex + " " + ey); }
  public final void drawRegularPolygon(int x, int y, int sides, int radius) { __cmds.add("polygon " + x + " " + y + " " + sides + " " + radius); }
  public final void drawEllipse(int x, int y, int width, int height) { __cmds.add("ellipse " + x + " " + y + " " + width + " " + height); }
  public final void drawRectangle(int x, int y, int width, int height) { __cmds.add("rectangle " + x + " " + y + " " + width + " " + height); }
  public final void drawShape(int[] points, boolean close) {
    String s = "shape";
    for (int p : points) s += " " + p;
    __cmds.add(s + " " + close);
  }
  public final void drawImage(Image image, int x, int y, int size) { __cmds.add("image " + image.getWidth() + "x" + image.getHeight() + " " + x + " " + y + " " + size); }
  public final void drawImage(Image image, int x, int y, int size, double rotation) { __cmds.add("image " + image.getWidth() + "x" + image.getHeight() + " " + x + " " + y + " " + size + " " + rotation); }
  public final void drawImage(Image image, int x, int y, int width, int height, double rotation) { __cmds.add("image " + image.getWidth() + "x" + image.getHeight() + " " + x + " " + y + " " + width + " " + height + " " + rotation); }
  public final void drawImage(String filename, int x, int y, int size) { __cmds.add("image " + filename + " " + x + " " + y + " " + size); }
  public final void drawImage(String filename, int x, int y, int size, double rotation) { __cmds.add("image " + filename + " " + x + " " + y + " " + size + " " + rotation); }
  public final void drawImage(String filename, int x, int y, int width, int height, double rotation) { __cmds.add("image " + filename + " " + x + " " + y + " " + width + " " + height + " " + rotation); }
}

class SoundLoader {
  // Audio decoding needs real assets (Phase 2); return silence so
  // sound levels compile and run headless.
  public static double[] read(String filename) { return new double[0]; }
}

class Theater {
  static void __write(Scene[] scenes) {
    try {
      java.io.PrintWriter w = new java.io.PrintWriter(new java.io.File("theater.log"));
      for (Scene s : scenes) for (String c : s.__commands()) w.println(c);
      w.close();
    } catch (Exception e) {}
  }
  public static void playScenes(Scene a) { __write(new Scene[]{a}); }
  public static void playScenes(Scene a, Scene b) { __write(new Scene[]{a, b}); }
  public static void playScenes(Scene a, Scene b, Scene c) { __write(new Scene[]{a, b, c}); }
  public static void playScenes(Scene a, Scene b, Scene c, Scene d) { __write(new Scene[]{a, b, c, d}); }
  public static void play(Scene[] scenes) { __write(scenes); }
}

File: apps/playground/src/theater.ts
/**
 * Theater visualizer: replays the `org.code.theater` draw-command log caturra
 * emits onto a 400×400 canvas — the browser-side equivalent of the real
 * javabuilder, which rasterizes the same commands to a GIF server-side with
 * `GraphicsHelper`/`GifWriter`.
 *
 * The bundled Scene records one command per line to `theater.log`; this module
 * parses and draws them, holding a frame at each `pause` for animation.
 */

const CANVAS = 400;
const SAMPLE_RATE = 44100;

/** Milliseconds a single `pause` frame holds, scaled to keep runs short. */
function stepMillis(pauseCount: number): number {
  return Math.min(400, Math.max(30, Math.floor(12_000 / Math.max(pauseCount, 1))));
}

export class TheaterViz {
  #ctx: CanvasRenderingContext2D;
  #run = 0;

  // Persisted drawing state (matches the real Scene's defaults).
  #fill: string | null = 'rgb(0,0,0)';
  #stroke: string | null = 'rgb(0,0,0)';
  #strokeWidth = 1;
  #textColor = 'rgb(0,0,0)';
  #textHeight = 20;
  #fontFamily = 'sans-serif';
  #fontStyle = '';

  // Web Audio for playNote, created lazily on the first note (after the Run
  // gesture, so autoplay policy is satisfied). Notes run through a per-run
  // master gain, so starting a new run silences whatever is still ringing.
  #audio: AudioContext | null = null;
  #master: GainNode | null = null;
  // Decoded bundled assets for playSound(name), and the current run's
  // playSound(double[]) sample buffers keyed by the log's pcm id.
  #assetBuffers = new Map<string, AudioBuffer>();
  #pcm = new Map<number, Float32Array>();
  // Decoded starter images for drawImage(name, …); names with no asset fall
  // back to the placeholder box.
  #images = new Map<string, ImageBitmap>();

  constructor(private readonly canvas: HTMLCanvasElement) {
    const ctx = canvas.getContext('2d');
    if (!ctx) {
      throw new Error('2d canvas context unavailable');
    }
    this.#ctx = ctx;
    canvas.width = CANVAS;
    canvas.height = CANVAS;
    this.reset();
  }

  /** Clear the canvas and reset drawing state to the Scene defaults. */
  reset(): void {
    this.#run += 1;
    this.silenceAudio();
    this.#fill = 'rgb(0,0,0)';
    this.#stroke = 'rgb(0,0,0)';
    this.#strokeWidth = 1;
    this.#textColor = 'rgb(0,0,0)';
    this.#textHeight = 20;
    this.#fontFamily = 'sans-serif';
    this.#fontStyle = '';
    this.#ctx.fillStyle = '#ffffff';
    this.#ctx.fillRect(0, 0, CANVAS, CANVAS);
  }

  /**
   * Replay a command log, animating pauses and playing sound. Cancels any
   * prior run. `pcm` maps each `sound pcm <id>` command to its samples (read
   * from the engine's VFS by the caller).
   */
  async play(log: string, pcm = new Map<number, Float32Array>()): Promise<void> {
    this.reset();
    this.#pcm = pcm;
    const run = this.#run;
    const commands = log.split('\n').filter((line) => line.length > 0);
    const pauseCount = commands.filter((c) => c.startsWith('pause ')).length;
    const hold = stepMillis(pauseCount);
    for (const command of commands) {
      if (this.#run !== run) {
        return; // superseded by a newer run
      }
      if (command.startsWith('pause ')) {
        await new Promise((resolve) => setTimeout(resolve, hold));
      } else {
        this.apply(command, true);
      }
    }
  }

  /** Apply every command instantly (a static final frame, for tests). */
  render(log: string): void {
    this.reset();
    for (const command of log.split('\n')) {
      if (command.length > 0 && !command.startsWith('pause ')) {
        this.apply(command, false);
      }
    }
  }

  private color(tokens: string[]): string | null {
    if (tokens[0] === 'none') {
      return null;
    }
    return `rgb(${tokens[0] ?? '0'},${tokens[1] ?? '0'},${tokens[2] ?? '0'})`;
  }

  private apply(command: string, audible: boolean): void {
    const ctx = this.#ctx;
    // `text "..." x y rot` needs the quoted body kept intact.
    const textMatch = /^text "(.*)" (-?\d+) (-?\d+) (-?[\d.]+)$/.exec(command);
    if (textMatch) {
      this.drawText(
        textMatch[1] ?? '',
        Number(textMatch[2]),
        Number(textMatch[3]),
        Number(textMatch[4]),
      );
      return;
    }
    const [op, ...rest] = command.split(' ');
    const n = (i: number): number => Number(rest[i] ?? 0);
    switch (op) {
      case 'clear':
        ctx.fillStyle = this.color(rest) ?? '#ffffff';
        ctx.fillRect(0, 0, CANVAS, CANVAS);
        break;
      case 'fillColor':
        this.#fill = this.color(rest);
        break;
      case 'strokeColor':
        this.#stroke = this.color(rest);
        break;
      case 'textColor':
        this.#textColor = this.color(rest) ?? 'rgb(0,0,0)';
        break;
      case 'strokeWidth':
        this.#strokeWidth = n(0);
        break;
      case 'textHeight':
        this.#textHeight = n(0);
        break;
      case 'textStyle':
        this.#fontFamily = rest[0] === 'MONO' ? 'monospace' : 'sans-serif';
        this.#fontStyle = rest[1] === 'BOLD' ? 'bold' : rest[1] === 'ITALIC' ? 'italic' : '';
        break;
      case 'rectangle':
        this.fillThenStroke(() => {
          ctx.rect(n(0), n(1), n(2), n(3));
        });
        break;
      case 'ellipse': {
        const [x, y, w, h] = [n(0), n(1), n(2), n(3)];
        this.fillThenStroke(() => {
          ctx.ellipse(x + w / 2, y + h / 2, w / 2, h / 2, 0, 0, Math.PI * 2);
        });
        break;
      }
      case 'line':
        if (this.#stroke) {
          ctx.strokeStyle = this.#stroke;
          ctx.lineWidth = this.#strokeWidth;
          ctx.beginPath();
          ctx.moveTo(n(0), n(1));
          ctx.lineTo(n(2), n(3));
          ctx.stroke();
        }
        break;
      case 'polygon': {
        const [x, y, sides, radius] = [n(0), n(1), n(2), n(3)];
        this.fillThenStroke(() => {
          const theta = (2 * Math.PI) / sides;
          for (let i = 0; i < sides; i++) {
            const px = Math.cos(theta * i) * radius + x;
            const py = Math.sin(theta * i) * radius + y;
            if (i === 0) ctx.moveTo(px, py);
            else ctx.lineTo(px, py);
          }
          ctx.closePath();
        });
        break;
      }
      case 'shape':
        this.drawShape(rest);
        break;
      case 'image':
        this.drawImage(rest);
        break;
      case 'note':
        if (audible) {
          this.playTone(rest);
        }
        break;
      case 'sound':
        if (audible) {
          this.playSample(rest);
        }
        break;
      default:
        break;
    }
  }

  private fillThenStroke(path: () => void): void {
    const ctx = this.#ctx;
    ctx.beginPath();
    path();
    if (this.#fill) {
      ctx.fillStyle = this.#fill;
      ctx.fill();
    }
    if (this.#stroke) {
      ctx.strokeStyle = this.#stroke;
      ctx.lineWidth = this.#strokeWidth;
      ctx.stroke();
    }
  }

  private drawText(text: string, x: number, y: number, rotation: number): void {
    const ctx = this.#ctx;
    ctx.save();
    if (rotation !== 0) {
      ctx.translate(x, y);
      ctx.rotate((rotation * Math.PI) / 180);
      ctx.translate(-x, -y);
    }
    ctx.font = `${this.#fontStyle} ${String(this.#textHeight)}px ${this.#fontFamily}`.trim();
    ctx.fillStyle = this.#textColor;
    ctx.textBaseline = 'alphabetic';
    ctx.textAlign = 'left';
    ctx.fillText(text, x, y);
    ctx.restore();
  }

  private drawShape(rest: string[]): void {
    const close = rest[rest.length - 1] === 'true';
    const points = rest.slice(0, -1).map(Number);
    const ctx = this.#ctx;
    if (close) {
      this.fillThenStroke(() => {
        for (let i = 0; i + 1 < points.length; i += 2) {
          const px = points[i] ?? 0;
          const py = points[i + 1] ?? 0;
          if (i === 0) ctx.moveTo(px, py);
          else ctx.lineTo(px, py);
        }
        ctx.closePath();
      });
    } else if (this.#stroke) {
      ctx.strokeStyle = this.#stroke;
      ctx.lineWidth = this.#strokeWidth;
      ctx.beginPath();
      for (let i = 0; i + 1 < points.length; i += 2) {
        const px = points[i] ?? 0;
        const py = points[i + 1] ?? 0;
        if (i === 0) ctx.moveTo(px, py);
        else ctx.lineTo(px, py);
      }
      ctx.stroke();
    }
  }

  /**
   * `image <name|WxH> x y size [rot]` or `image <name|WxH> x y w h rot`. A named
   * starter asset draws for real; an in-engine `Image` object (logged as its
   * dimensions, with no pixels) still draws the placeholder box.
   */
  private drawImage(rest: string[]): void {
    const key = rest[0] ?? '';
    const bitmap = this.#images.get(key);
    if (!bitmap) {
      this.drawImagePlaceholder(rest);
      return;
    }
    const nums = rest.slice(1).map(Number);
    const [x, y] = [nums[0] ?? 0, nums[1] ?? 0];
    let width: number;
    let height: number;
    let rotation: number;
    if (nums.length >= 5) {
      // x y w h rot
      width = nums[2] ?? bitmap.width;
      height = nums[3] ?? bitmap.height;
      rotation = nums[4] ?? 0;
    } else {
      // x y size [rot] — `size` is the width; the height keeps the aspect ratio.
      width = nums[2] ?? bitmap.width;
      height = bitmap.width === 0 ? width : (width * bitmap.height) / bitmap.width;
      rotation = nums[3] ?? 0;
    }
    const ctx = this.#ctx;
    ctx.save();
    if (rotation !== 0) {
      ctx.translate(x, y);
      ctx.rotate((rotation * Math.PI) / 180);
      ctx.translate(-x, -y);
    }
    ctx.drawImage(bitmap, x, y, width, height);
    ctx.restore();
  }

  /** No asset for this name (or an in-engine Image): a labelled placeholder box. */
  private drawImagePlaceholder(rest: string[]): void {
    const nums = rest.slice(1).map(Number);
    const [x, y, size] = [nums[0] ?? 0, nums[1] ?? 0, nums[2] ?? 40];
    const ctx = this.#ctx;
    ctx.fillStyle = '#dcdce4';
    ctx.strokeStyle = '#9a9aa8';
    ctx.lineWidth = 1;
    ctx.fillRect(x, y, size, size);
    ctx.strokeRect(x, y, size, size);
  }

  /** The shared AudioContext, created on first use; null where Web Audio is absent. */
  private audioContext(): AudioContext | null {
    if (this.#audio) {
      return this.#audio;
    }
    if (typeof AudioContext === 'undefined') {
      return null;
    }
    this.#audio = new AudioContext();
    return this.#audio;
  }

  /**
   * Fetch and decode named assets (`name` → URL), returning their mono samples
   * by name so the caller can preload them into the engine's VFS for
   * `SoundLoader.read`. Decoded buffers are cached and reused to play
   * `sound file <name>`. Whatever the stored format, decodeAudioData resamples
   * to the AudioContext's rate, so the program still sees 44.1 kHz samples.
   */
  async loadAssets(assets: Map<string, string>): Promise<Map<string, Float32Array>> {
    const ctx = this.audioContext();
    const samples = new Map<string, Float32Array>();
    if (!ctx) {
      return samples;
    }
    for (const [name, url] of assets) {
      let buffer = this.#assetBuffers.get(name);
      if (!buffer) {
        try {
          const response = await fetch(url);
          buffer = await ctx.decodeAudioData(await response.arrayBuffer());
          this.#assetBuffers.set(name, buffer);
        } catch {
          continue; // asset unavailable — SoundLoader.read falls back to silence
        }
      }
      samples.set(name, buffer.getChannelData(0));
    }
    return samples;
  }

  /** Fetch and decode named images (`name` → URL) for `drawImage(name, …)`. */
  async loadImages(assets: Map<string, string>): Promise<void> {
    for (const [name, url] of assets) {
      if (this.#images.has(name)) {
        continue;
      }
      try {
        const response = await fetch(url);
        this.#images.set(name, await createImageBitmap(await response.blob()));
      } catch {
        // Unavailable — drawImage falls back to the placeholder box.
      }
    }
  }

  /** Play a `sound file <name>` (bundled asset) or `sound pcm <id> <len>` command. */
  private playSample(rest: string[]): void {
    const ctx = this.audioContext();
    if (!ctx) {
      return;
    }
    if (ctx.state === 'suspended') {
      void ctx.resume();
    }
    let buffer: AudioBuffer | undefined;
    if (rest[0] === 'file') {
      buffer = this.#assetBuffers.get(rest[1] ?? '');
    } else if (rest[0] === 'pcm') {
      const samples = this.#pcm.get(Number(rest[1]));
      if (samples && samples.length > 0) {
        buffer = ctx.createBuffer(1, samples.length, SAMPLE_RATE);
        // Copy into a fresh ArrayBuffer-backed view (createBuffer's channel type).
        buffer.copyToChannel(new Float32Array(samples), 0);
      }
    }
    if (!buffer) {
      return;
    }
    const source = ctx.createBufferSource();
    source.buffer = buffer;
    source.connect(this.masterGain(ctx));
    source.start(ctx.currentTime);
  }

  /** The current run's master gain, created lazily and rebuilt each run. */
  private masterGain(ctx: AudioContext): GainNode {
    if (!this.#master) {
      const gain = ctx.createGain();
      gain.gain.value = 1;
      gain.connect(ctx.destination);
      this.#master = gain;
    }
    return this.#master;
  }

  /** Fade out notes still ringing from the previous run; next run gets a fresh master. */
  private silenceAudio(): void {
    if (this.#master && this.#audio) {
      const now = this.#audio.currentTime;
      this.#master.gain.cancelScheduledValues(now);
      this.#master.gain.setTargetAtTime(0, now, 0.02);
    }
    this.#master = null;
  }

  /**
   * Synthesize a `note <INSTRUMENT> <midi> <seconds>` command with a single
   * oscillator and a percussive envelope. MIDI note 69 (A4) is 440 Hz;
   * PIANO is a triangle wave, BASS a sawtooth. The audible length is clamped
   * because the visual timeline is compressed — a long nominal note shouldn't
   * ring over the rest of the scene.
   */
  private playTone(rest: string[]): void {
    const ctx = this.audioContext();
    if (!ctx) {
      return;
    }
    if (ctx.state === 'suspended') {
      void ctx.resume();
    }
    const midi = Number(rest[1] ?? 60);
    const seconds = Number(rest[2] ?? 0.5);
    if (!Number.isFinite(midi)) {
      return;
    }
    const bass = rest[0] === 'BASS';
    const frequency = 440 * 2 ** ((midi - 69) / 12);
    const duration = Math.min(Math.max(Number.isFinite(seconds) ? seconds : 0.5, 0.12), 1.5);

    const osc = ctx.createOscillator();
    osc.type = bass ? 'sawtooth' : 'triangle';
    osc.frequency.value = frequency;

    const gain = ctx.createGain();
    const t0 = ctx.currentTime;
    const peak = bass ? 0.3 : 0.22;
    gain.gain.setValueAtTime(0.0001, t0);
    gain.gain.exponentialRampToValueAtTime(peak, t0 + 0.01);
    gain.gain.exponentialRampToValueAtTime(0.0001, t0 + duration);

    osc.connect(gain);
    gain.connect(this.masterGain(ctx));
    osc.start(t0);
    osc.stop(t0 + duration + 0.02);
  }
}
